Source/Diagnostics/ReducedDiags/Costs.cpp

```diff
@@ -0,0 +1,199 @@
+/* Copyright 2019-2020 Yinjian Zhao
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "Costs.H"
+#include "WarpX.H"
+
+
+using namespace amrex;
+
+// constructor
+Costs::Costs (std::string rd_name)
+    : ReducedDiags{rd_name}
+{
+
+}
+
+// function that gathers costs
+void Costs::ComputeDiags (int step)
+{
+    // get WarpX class object
+    auto& warpx = WarpX::GetInstance();
+
+    // for now, costs reduced diagnostic only works with `costs_heuristic`,
+    // but this will work for timer based costs as well after changing from
+    // multifab to vector
+    const amrex::Vector<amrex::Real>* cost_heuristic = warpx.getCostsHeuristic(0);
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(cost_heuristic != nullptr,
+        "Costs reduced diagnostic does not work with timer-based costs update.");
+
+    // judge if the diags should be done
+    // costs is initialized only if we're doing load balance
+    if ( ((step+1) % m_freq != 0) || warpx.get_load_balance_int() < 1 ) { return; }
+
+    // get number of boxes over all levels
+    auto nLevels = warpx.finestLevel() + 1;
+    int nBoxes = 0;
+    for (int lev = 0; lev < nLevels; ++lev)
+    {
+        const amrex::Vector<amrex::Real>* cost_heuristic = warpx.getCostsHeuristic(lev);
+        nBoxes += cost_heuristic->size();
+    }
+
+    // keep track of the max number of boxes, this is needed later on to fill
+    // the jagged array (in case each step does not have the same number of boxes)
+    nBoxesMax = std::max(nBoxesMax, nBoxes);
+
+    // resize and clear data array
+    m_data.resize(nDataFields*nBoxes, 0.0);
+    m_data.assign(nDataFields*nBoxes, 0.0);
+
+    // if not a load balance step, we must compute costs
+    if ((step+1) % warpx.get_load_balance_int() != 0)
+    {
+        if (warpx.load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::Heuristic)
+        {
+            warpx.ComputeCostsHeuristic();
+
+        } else if (WarpX::load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::Timers)
+        {
+            // costs update via timers, nothing to compute
+        }
```

Comment:
If `costs update via timers, nothing to compute`, do we still need this `else if`? Can't a single `if (warpx.load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::Heuristic)` be enough?

Reply:
Good point, I have condensed this.